Xamarin.PropertyEditing/ObservableLookup.cs

```diff
@@ -0,0 +1,336 @@
+				group = this.nullGrouping;
+			} else {
+				if (!this.groupings.TryGetValue (key, out group))
+					return false;
+			}
+
+			if (group.Remove (element)) {
+				if (group.Count == 0) {
+					Remove (key);
+				}
+
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// Removes <paramref name="key"/> from the lookup.
+		/// </summary>
+		/// <param name="key">They to remove.</param>
+		/// <returns><c>true</c> if <paramref name="key"/> existed.</returns>
+		/// <exception cref="ArgumentNullException"><paramref name="key"/> is <c>null</c>.</exception>
+		public bool Remove (TKey key)
+		{
+			if (key == null) {
+				bool removed = (this.nullGrouping.Count > 0);
+				this.nullGrouping.Clear ();
+				if (removed)
+					OnCollectionChanged (new NotifyCollectionChangedEventArgs (NotifyCollectionChangedAction.Remove, this.nullGrouping, 0));
+
+				return removed;
+			}
+
+			int index = this.groupings.IndexOf (key);
+			if (index >= 0) {
+				var g = this.groupings[index];
+				this.groupings.Remove (key);
+				if (ReuseGroups) {
+					g.QuietClear ();
+					this.groupings.Add (key, g);
+				}
+				
+				OnCollectionChanged (new NotifyCollectionChangedEventArgs (NotifyCollectionChangedAction.Remove, g, index));
+				return true;
+			}
+
+			return false;
+		}
+
+		public void Clear ()
+		{
+			if (this.nullGrouping != null)
+				this.nullGrouping.Clear ();
+
+			if (ReuseGroups) {
+				foreach (var g in this.groupings.Values) {
+					this.oldGroups.Add (g.Key, g);
+					g.QuietClear ();
+				}
+			}
+
+			this.groupings.Clear ();
+			OnCollectionChanged (new NotifyCollectionChangedEventArgs (NotifyCollectionChangedAction.Reset));
+		}
+
+		public bool TryGetValues (TKey key, out IEnumerable<TElement> values)
+		{
+			values = null;
+
+			if (key == null) {
+				if (this.nullGrouping.Count != 0) {
+					values = nullGrouping;
+					return true;
+				} else
+					return false;
+			}
+
+			ObservableGrouping<TKey, TElement> grouping;
+			if (!this.groupings.TryGetValue (key, out grouping))
+				return false;
+
+			values = grouping;
+			return true;
+		}
+
+		#region ILookup Members
+		/// <summary>
+		/// Gets the number of groupings.
+		/// </summary>
+		public int Count
+		{
+			get { return this.groupings.Count + ((this.nullGrouping != null && this.nullGrouping.Count > 0) ? 1 : 0); }
+		}
+
+		IGroupingList<TKey, TElement> IReadOnlyList<IGroupingList<TKey, TElement>>.this[int index] => (IGroupingList<TKey, TElement>)this[index];
+
+		/// <summary>
+		/// Gets the elements for <paramref name="key"/>.
+		/// </summary>
+		/// <param name="key">The key to get the elements for.</param>
+		/// <returns>The elements under <paramref name="key"/>.</returns>
+		public IEnumerable<TElement> this[TKey key]
+		{
+			get
+			{
+				if (key == null)
+					return this.nullGrouping;
+
+				ObservableGrouping<TKey, TElement> grouping;
+				if (this.groupings.TryGetValue (key, out grouping))
+					return grouping;
+
+				return new TElement[0];
```

Comment:
`Enumerable.Empty<TElement>` or `Array.Empty<TElement>` should be preferred generally